File: articles/broaden-and-build-theory-positive-psychology/c/broaden_build_index.c
#include <stdio.h>

typedef struct {
    double positive_emotion;
    double negative_emotion;
    double cognitive_flexibility;
    double exploratory_behavior;
    double affiliative_behavior;
    double social_support;
    double resilience_score;
    double stress_arousal;
    double contextual_safety;
    double resource_stock;
    double practice_fit;
} BroadenBuildIndicators;

double broadening_index(BroadenBuildIndicators x) {
    return (
        x.positive_emotion +
        x.cognitive_flexibility +
        x.exploratory_behavior +
        x.affiliative_behavior +
        x.contextual_safety -
        x.negative_emotion
    ) / 6.0;
}

double resource_index(BroadenBuildIndicators x) {
    return (
        x.social_support +
        x.resilience_score +
        x.contextual_safety +
        x.resource_stock +
        x.practice_fit
    ) / 5.0;
}

double recovery_capacity_index(BroadenBuildIndicators x) {
    return (
        x.positive_emotion +
        x.social_support +
        x.contextual_safety +
        x.resilience_score -
        x.stress_arousal -
        x.negative_emotion
    ) / 6.0;
}

double net_adaptation_index(BroadenBuildIndicators x) {
    return x.positive_emotion
         + x.cognitive_flexibility
         + x.exploratory_behavior
         + x.affiliative_behavior
         + x.social_support
         + x.resilience_score
         + x.contextual_safety
         + x.resource_stock
         + x.practice_fit
         - x.negative_emotion
         - x.stress_arousal;
}

int main(void) {
    BroadenBuildIndicators example = {7.2, 3.8, 7.1, 7.0, 7.2, 7.3, 6.9, 3.6, 7.2, 7.1, 7.3};
    printf("Broadening index: %.3f\n", broadening_index(example));
    printf("Resource index: %.3f\n", resource_index(example));
    printf("Recovery capacity index: %.3f\n", recovery_capacity_index(example));
    printf("Net adaptation index: %.3f\n", net_adaptation_index(example));
    return 0;
}

File: articles/broaden-and-build-theory-positive-psychology/rust/broaden_build_index.rs
struct BroadenBuildIndicators {
    positive_emotion: f64,
    negative_emotion: f64,
    cognitive_flexibility: f64,
    exploratory_behavior: f64,
    affiliative_behavior: f64,
    social_support: f64,
    resilience_score: f64,
    stress_arousal: f64,
    contextual_safety: f64,
    resource_stock: f64,
    practice_fit: f64,
}

fn broadening_index(x: &BroadenBuildIndicators) -> f64 {
    (x.positive_emotion
        + x.cognitive_flexibility
        + x.exploratory_behavior
        + x.affiliative_behavior
        + x.contextual_safety
        - x.negative_emotion)
        / 6.0
}

fn resource_index(x: &BroadenBuildIndicators) -> f64 {
    (x.social_support + x.resilience_score + x.contextual_safety + x.resource_stock + x.practice_fit) / 5.0
}

fn recovery_capacity_index(x: &BroadenBuildIndicators) -> f64 {
    (x.positive_emotion
        + x.social_support
        + x.contextual_safety
        + x.resilience_score
        - x.stress_arousal
        - x.negative_emotion)
        / 6.0
}

fn net_adaptation_index(x: &BroadenBuildIndicators) -> f64 {
    x.positive_emotion
        + x.cognitive_flexibility
        + x.exploratory_behavior
        + x.affiliative_behavior
        + x.social_support
        + x.resilience_score
        + x.contextual_safety
        + x.resource_stock
        + x.practice_fit
        - x.negative_emotion
        - x.stress_arousal
}

fn main() {
    let example = BroadenBuildIndicators {
        positive_emotion: 7.2,
        negative_emotion: 3.8,
        cognitive_flexibility: 7.1,
        exploratory_behavior: 7.0,
        affiliative_behavior: 7.2,
        social_support: 7.3,
        resilience_score: 6.9,
        stress_arousal: 3.6,
        contextual_safety: 7.2,
        resource_stock: 7.1,
        practice_fit: 7.3,
    };

    println!("Broadening index: {:.3}", broadening_index(&example));
    println!("Resource index: {:.3}", resource_index(&example));
    println!("Recovery capacity index: {:.3}", recovery_capacity_index(&example));
    println!("Net adaptation index: {:.3}", net_adaptation_index(&example));
}
